prov/shm/src/smr_dsa.c

```diff
@@ -0,0 +1,868 @@
+			if (wstate != ACCFG_WQ_ENABLED)
+				continue;
+
+			if (accfg_wq_get_max_transfer_size(wq) < SMR_SAR_SIZE)
+				continue;
+
+			/* The wq type should be user */
+			type = accfg_wq_get_type(wq);
+			if (type != ACCFG_WQT_USER)
+				continue;
+
+			/* Make sure the mode is correct */
+			int mode = accfg_wq_get_mode(wq);
+			if ((mode == ACCFG_WQ_SHARED && !shared) ||
+			    (mode == ACCFG_WQ_DEDICATED && shared))
+				continue;
+
+			/* This is a candidate wq */
+			wq_reg = idxd_wq_mmap(wq);
+			if (wq_reg == NULL)
+				continue;
+			wq_array[wq_count] = wq_reg;
+			wq_count++;
+			break;
+		}
+
+		if (wq_count >= MAX_WQS_PER_EP)
+			break;
+	}
+
+	accfg_unref(ctx);
+	return wq_count;
+}
+
+static int bitmap_test_and_set_bit(struct dsa_bitmap *bitmap, int index)
+{
+	assert(index < bitmap->size);
+	return atomic_fetch_or(&bitmap->data, 1ULL << index) & (1ULL << index);
+}
+
+static void bitmap_clear_bit(struct dsa_bitmap *bitmap, int index)
+{
+	assert(index < bitmap->size);
+	atomic_fetch_and(&bitmap->data, ~(1ULL << index));
+}
+
+static int bitmap_allocate(struct dsa_bitmap *bitmap, int size)
+{
+	assert(size <= MAX_BITMAP_SIZE);
+
+	atomic_init(&bitmap->data, 0);
+	bitmap->size = size;
+
+	return 1;
+}
+
+static int bitmap_test_bit(struct dsa_bitmap *bitmap, int index)
+{
+	assert(index < bitmap->size);
+	return atomic_load(&bitmap->data) & (1ULL << index);
+}
+
+static int bitmap_is_empty(struct dsa_bitmap *bitmap)
+{
+	return atomic_load(&bitmap->data) == 0;
+}
+
+/* DSA_CMD_CONTEXT - START */
+static struct dsa_cmd_context *
+allocate_dsa_cmd_context(struct smr_dsa_context *smr_dsa_context)
+{
+	struct dsa_cmd_context *dsa_cmd_context;
+	int i;
+
+	for (i = 0; i < DSA_CMD_CONTEXT_COUNT; i++) {
+		if (!bitmap_test_and_set_bit(&smr_dsa_context->dsa_bitmap, i))
+			break;
+	}
+
+	if (i == DSA_CMD_CONTEXT_COUNT)
+		return NULL;
+
+	dsa_cmd_context = &smr_dsa_context->dsa_cmd_context[i];
+	memset(dsa_cmd_context, 0, sizeof(struct dsa_cmd_context));
+	dsa_cmd_context->index = i;
+
+	return dsa_cmd_context;
+}
+
+static void free_dsa_cmd_context(struct dsa_cmd_context *dsa_cmd_context,
+			  struct smr_dsa_context *smr_dsa_context)
+{
+	bitmap_clear_bit(&smr_dsa_context->dsa_bitmap, dsa_cmd_context->index);
+}
+
+static struct dsa_hw_desc *
+get_work_descriptor_array_ptr(struct dsa_cmd_context *dsa_cmd_context,
+			      struct smr_dsa_context *dsa_context)
+{
+	return &dsa_context->dsa_work_desc[dsa_cmd_context->index *
+					   MAX_CMD_BATCH_SIZE];
+}
+
+static struct dsa_hw_desc *
+get_free_work_descriptor(struct dsa_cmd_context *dsa_cmd_context,
+			 struct smr_dsa_context *dsa_context)
+{
+	struct dsa_hw_desc *free_desc =
+	    &dsa_context
+		 ->dsa_work_desc[dsa_cmd_context->index * MAX_CMD_BATCH_SIZE +
+				 dsa_cmd_context->batch_size];
+	struct dsa_completion_record *free_comp =
+	    &dsa_context
+		 ->dsa_work_comp[dsa_cmd_context->index * MAX_CMD_BATCH_SIZE +
+				 dsa_cmd_context->batch_size++];
```

Comment:
The above lines are, like, hard to read.  Please separate declaration from assignment and fixup indentation.